### PR TITLE
Article css filtering

### LINKED ISSUE #3046: Idea - exclude certain CSS elements from page
Excellent programme - I use it all the time to pull full web pages using CSS selector. However, sometimes in the page there are elements I don't want to see, such as newsletter sign-up boxes, 'Most read' etc, and these have their own CSS selector which is consistent within the page. Is it possible to list elements which I don't want shown and for those elements to be dumped.

In the screenshot I've put a red box around two areas, the top one is an email signup form, the bottom one is 'related articles' - both of these appear in every article, and are both inside the article body.
Screenshot_2020-06-09_23-48-05

### LINKED ISSUE #3406: how do I create myself theme
The default theme is nice, but I want to create myself thems. 
Is there any way to get?

### PR DESCRIPTION
**Pull request by papaschloss**
lets you add comma separated CSS selectors that will be removed from the extracted article content. Preview, import, and export work too. Let me know what else I can do to make this PR better

How to test the feature manually:

1. Add a feed with CSS selector and filtering to make sure the filtered elements are removed
2. Add a feed with CSS selector and no filtering to make sure plain old filtering works
3. Preview a feed via the css selector to make sure the css filters are removed in the preview
4. Export and import to make sure filtering OPML elements work correctly

Pull request checklist:

- [ ] clear commit messages
- [ ] code manually tested
- [ ] unit tests written (optional if too hard)
- [ ] documentation updated

### CODE REVIEW

File: app/SQL/install.sql.mysql.php [hunk 1]
```diff
@@ -27,6 +27,7 @@
 	`lastUpdate` INT(11) DEFAULT 0,	-- Until year 2038
 	`priority` TINYINT(2) NOT NULL DEFAULT 10,
 	`pathEntries` VARCHAR(511) DEFAULT NULL,
+	`pathEntriesFilter` VARCHAR(511) DEFAULT NULL,
```
Comment (Alkarex) [inline app/SQL/install.sql.mysql.php]: Please save this information in the JSON `attributes` column instead. I would rather avoid a new column for this type of data. See example of `xpath`

**Comment by Alkarex** [comment]
Thanks and welcome! 👍🏻

**Comment by Alkarex** [comment]
Please run `make fix-all` and `make test-all` if you can

**Comment by papaschloss** [comment]
Updated to use the json attributes storage instead of DB column. These changes fix the fix-all and test-all issues too.

**Comment by math-GH** [comment]
I do not understand it :/

How is it related to #3406 (it is about creating a new theme)?

File: app/Models/DatabaseDAO.php [hunk 2]
```diff
@@ -89,7 +89,7 @@ public function categoryIsCorrect(): bool {
 	public function feedIsCorrect(): bool {
 		return $this->checkTable('feed', array(
 			'id', 'url', 'category', 'name', 'website', 'description', 'lastUpdate',
-			'priority', 'pathEntries', 'httpAuth', 'error', 'ttl', 'attributes',
+			'priority', 'pathEntries', 'pathEntriesFilter', 'httpAuth', 'error', 'ttl', 'attributes',
```
Comment (Alkarex) [inline app/Models/DatabaseDAO.php]: This line is not needed anymore, is it?

File: app/Models/Entry.php [hunk 3]
```diff
@@ -558,6 +558,12 @@ public static function getContentByParsing(string $url, string $path, array $att
 			$nodes = $xpath->query(new Gt\CssXPath\Translator($path));
 			if ($nodes != false) {
 				foreach ($nodes as $node) {
+					if ($attributes['path_entries_filter'] != '') {
```
Comment (Alkarex) [inline app/Models/Entry.php]:
```suggestion
					if (!empty($attributes['path_entries_filter'])) {
```

More robust if this array entry does not exist at all

File: app/i18n/fr/sub.php [hunk 4]
```diff
@@ -57,6 +57,7 @@
 		'css_cookie_help' => 'Exemple : <kbd>foo=bar; gdpr_consent=true; cookie=value</kbd>',
 		'css_help' => 'Permet de récupérer les flux tronqués (attention, demande plus de temps !)',
 		'css_path' => 'Sélecteur CSS des articles sur le site d’origine',
+		'css_path_filter' => 'Article CSS selectors to remove, comma separated',	// TODO
```
Comment (Alkarex) [inline app/i18n/fr/sub.php]:
```suggestion
		'css_path_filter' => 'Sélecteurs CSS des éléments à supprimer, séparés par une virgule',
```

File: p/scripts/feed.js [hunk 5]
```diff
@@ -56,7 +56,8 @@ function init_popup_preview_selector() {
 
 	link.addEventListener('click', function (ev) {
 		const selector_entries = document.getElementById('path_entries').value;
-		const href = link.href.replace('selector-token', encodeURIComponent(selector_entries));
+		const selector_entries_filter = document.getElementById('path_entries_filter').value;
+		const href = link.href.replace('selector-token', encodeURIComponent(selector_entries)).replace('selector-filter-token', encodeURIComponent(selector_entries_filter));
```
Comment (Alkarex) [inline p/scripts/feed.js]:
```suggestion
		const href = link.href.replace('selector-token', encodeURIComponent(selector_entries)).
				replace('selector-filter-token', encodeURIComponent(selector_entries_filter));
```

File: app/Models/DatabaseDAO.php (hunk 2 above)
Comment (papaschloss) [reply app/Models/DatabaseDAO.php]: fixed

**Comment by papaschloss** [comment]
all issues are fixed

File: app/Controllers/feedController.php [hunk 6]
```diff
@@ -916,12 +916,15 @@ public function contentSelectorPreviewAction() {
 			return;
 		}
 
+		$attributes = $feed->attributes();
+		$attributes['path_entries_filter'] = trim(Minz_Request::param('selector_filter'));
```
Comment (Alkarex) [inline app/Controllers/feedController.php]:
```suggestion
		$attributes['path_entries_filter'] = trim(Minz_Request::param('selector_filter', ''));
```

trim() does not accept (anymore, PHP8.1+) a null, which is the default value

**Comment by Alkarex** [comment]
Thanks again! Please add a line for you in https://github.com/FreshRSS/FreshRSS/blob/edge/CREDITS.md

**Comment by nfriedly** [comment]
Thank you for adding this!

**Pull request merged**